WIP Adds facebox_face_detect

robmarkcole opened the pull request:
Replaces https://github.com/home-assistant/home-assistant/pull/14248

## Description:
Adds component for face detection (number of faces) using [facebox](https://machineboxio.com/docs/facebox/teaching-facebox). Run facebox with:
```
MB_KEY="INSERT-YOUR-KEY-HERE"

docker run -p 8080:8080 -e "MB_KEY=$MB_KEY" -e "MB_FACEBOX_DISABLE_RECOGNITION=true" machinebox/facebox
```

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io/pull/5300) with documentation (if applicable):** home-assistant/home-assistant.github.io#<5300>

## Example entry for `configuration.yaml` (if applicable):
```yaml
image_processing:
  - platform: facebox_face_detect
    ip_address: 192.168.0.1
    port: 8080
    source:
      - entity_id: camera.local_file
```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

dlbroadfoot committed:
Fix Gogogate2 'available' attribute (#13728)

* Fixed bug -  unable to set base readaonly property

* PR fixes

* Added line

Kane610 committed:
Fix so it is possible to ignore discovered config entry handlers (#13741)

* Fix so it is possible to ignore discovered config entry handlers

* Improve efficiency

balloob committed:
Prevent vesync doing I/O in event loop (#13862)

syssi committed:
Missing property decorator added (#13889)

balloob committed:
Fix race condition for component loaded before listening (#13887)

* Fix race condition for component loaded before listening

* async/await syntax

aav7fl committed:
Fix #13846 Double underscore in bluetooth address (#13884)

syssi committed:
Import operation modes from air humidifier (#13908)

kellerza committed:
Upgrade pyqwikswitch to 0.71 (#13920)

titilambert committed:
Update pyhydroquebec to 2.2.2 (#13946)

titilambert committed:
Update pyfido to 2.1.1 (#13947)

balloob committed:
Upgrade somecomfort to 0.5.2 (#13940)

Danielhiversen committed:
Fix call to parent broadlink switch (#13906)

* Broadlink switch, fixes issue #13799

* slugify

bachya committed:
Bumped pypollencom to 1.1.2 (#13959)

* Bumped pypollencom to 1.1.2

* Updated requirements_all.txt

balloob committed:
Version bump to 0.67.1

balloob committed:
Revert "Upgrade pyqwikswitch to 0.71 (#13920)"

This reverts commit 6fa60c464bb012c73f03b9b3ea69d82539d7c8f5.

balloob committed:
Merge pull request #13975 from home-assistant/rc

0.67.1

balloob committed:
Merge branch 'dev' into rc

balloob committed:
Version bump to 0.68.0b0

kellerza committed:
Qwikswitch binary sensors (#14008)

schmittx committed:
HomeKit Alarm Control Panel Code Exception Fix (#14025)

* Catch exception for KeyError
* Use get and added test

OttoWinter committed:
Revert cast platform polling mode (#14027)

mjg59 committed:
Handle HomeKit configuration failure more cleanly (#14041)

* Handle HomeKit configuration failure more cleanly

Add support for handling cases where HomeKit configuration fails, and give
the user more information about what to do.

* Don't consume the exception for a homekit.UnknownError

If we get an UnknownError then we should alert the user but also still
generate the backtrace so there's actually something for them to file in
a bug report.

marthoc committed:
Update device classes for contact sensor HomeKit (#14051)

balloob committed:
Order the output of the automation editor (#14019)

* Order the output of the automation editor

* Lint

Danielhiversen committed:
Upgrade broadlink lib (#14074)

balloob committed:
Bump frontend to 20180425.0

balloob committed:
Version bump to 0.68.0b1

Danielhiversen committed:
Fix timezone issue when calculating min/max values in tibber #14009 (#14080)

* fix timezone issue in tibber #14009

* remove debug print

GotoCode committed:
Updated list of AWS regions for Amazon Polly (#14097)

Fixes #14052

balloob committed:
Bump frontend to 20180426

balloob committed:
Version bump to 0.68.0b2

amelchio committed:
Improve precision of Hue color state (#14113)

balloob committed:
Version bump to 0.68.0

mjg59 committed:
Change Eufy brightness handling (#14111)

Eufy device state isn't reported if the bulb is off, so avoid stamping on
the previous values if the bulb isn't going to give us useful information.
In addition, improve handling of bulb turn on if we aren't provided with a
brightness - this should avoid the bulb tending to end up with a brightness of
1 after power cycling.

balloob committed:
Merge pull request #14124 from home-assistant/rc

0.68

amelchio committed:
Fix color setting of tplink lights (#14108)

balloob committed:
Disable eliqonline requirement (#14156)

* Disable eliqonline requirement

* Disable pylint import error

mjg59 committed:
Fix Python 3.6 compatibility for HomeKit controller (#14160)

Python 3.6's http client passes an additional argument to _send_output,
so add that to the function definition.

amelchio committed:
Revert Hue color state to be xy-based (#14154)

OttoWinter committed:
Improve chromecast disconnection logic (#14190)

* Attempt Cast Fix

* Cleanup

cdce8p committed:
Added CONF_IP_ADDRESS to HomeKit (#14163)

balloob committed:
Do not sync entities with an empty name (#14181)

balloob committed:
Fix poorly formatted automations (#14196)

balloob committed:
Version bump to 0.68.1

balloob committed:
Merge pull request #14199 from home-assistant/rc

0.68.1

robmarkcole committed:
Adds facebox face detect

robmarkcole committed:
Adds facebox face detect

robmarkcole committed:
Adds facebox face detect

robmarkcole committed:
Adds facebox face detect

Danielhiversen commented on homeassistant/components/image_processing/facebox_face_detect.py:
```diff
@@ -0,0 +1,107 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+ROUNDING_DECIMALS = 2
```
Unused?

robmarkcole replied:
True!

robmarkcole commented:
I'm getting an error in the test: 
```
entity.py                  206 ERROR    Update for image_processing.facebox_demo_camera fails
Traceback (most recent call last):
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/helpers/entity.py", line 204, in async_update_ha_state
    yield from self.async_device_update()
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/helpers/entity.py", line 325, in async_device_update
    yield from self.async_update()
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/components/image_processing/__init__.py", line 135, in async_update
    yield from self.async_process_image(image.content)
  File "/Users/robincole/miniconda3/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/components/image_processing/facebox_face_detect.py", line 70, in process_image
    timeout=30
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/requests/api.py", line 112, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/requests/api.py", line 58, in request
    return session.request(method=method, url=url, **kwargs)
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/requests/sessions.py", line 508, in request
    resp = self.send(prep, **send_kwargs)
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/requests_mock/mocker.py", line 104, in _fake_send
    return _original_send(session, request, **kwargs)
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/requests/sessions.py", line 618, in send
    r = adapter.send(request, **kwargs)
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/requests_mock/adapter.py", line 233, in send
    raise exceptions.NoMockAddress(request)
requests_mock.exceptions.NoMockAddress: No mock address: POST http://192.168.0.1:8080/facebox/check
```
Is this because I am not posting the file?

robmarkcole committed:
Remove ROUNDING_DECIMALS

robmarkcole committed:
Remove ROUNDING_DECIMALS

robmarkcole committed:
Remove ROUNDING_DECIMALS

fabaff committed:
Update issue templates

fabaff committed:
Update issue templates

fabaff committed:
Update issue templates

fabaff committed:
Update issue templates

fabaff committed:
Remove feature request

MartinHjelmare commented on homeassistant/components/image_processing/facebox_face_detect.py:
```diff
@@ -0,0 +1,105 @@
+    PLATFORM_SCHEMA, CONF_SOURCE, CONF_ENTITY_ID,
+    CONF_NAME)
+from homeassistant.const import (CONF_IP_ADDRESS, CONF_PORT)
+from homeassistant.components.image_processing.microsoft_face_identify import (
```
Why is the base face entity class located in the microsoft platform? This is weird. @pvizeli do you know? Normally we should not rely on another platform in a platform. The base entity class should probably move to `homeassistant.components.image_processing`.

robmarkcole replied:
@MartinHjelmare I agree and propose a refactor

pvizeli replied:
@MartinHjelmare that is correct. On my inital PR it was inside component but I need remove this. In a later discusion about this designe, paulus agree that I can move back the code. But I found never time for this.

MartinHjelmare replied:
#robmarkcole if you have time, a refactor in a separate PR would be great.

robmarkcole replied:
Sure

amelchio committed:
Color fixes for Wink lights (#14263)

fanthos committed:
fix check config not working after #14211 (#14259)

OttoWinter committed:
Issue Template Fix CRLF (#14283)

pvizeli commented:
Maybe you can convert this to async in future :)

pvizeli commented on homeassistant/components/image_processing/facebox_face_detect.py:
```diff
@@ -0,0 +1,105 @@
+
+        if response['success']:
+            elapsed_time = time.perf_counter() - timer_start
+            self._response_time = "{} seconds".format(
```
Please update the time only if you find things. Otherwise it flush the state database.

robmarkcole replied:
you mean if `total_faces` >  zero?

pvizeli replied:
yes. If there is no faces found, he should also not update time.

robmarkcole replied:
@pvizeli That would mean that the `response_time` was for an earlier (successful request), and not for the last successful request (that returned zero faces). I think the confusion that could cause outweighs any benefit from not updating? The only alternative I can think of is to drop `response_time`  but it is useful for choosing a `scan_interval` and monitoring performance of the classifier under different loads

fabaff committed:
Upgrade attrs to 18.1.0 (#14281)

cdce8p committed:
Update HAP-python to 2.0.0 (#14278)

* Fixed async (added 'async_add_job' and 'add_job')

* Driver status

* Use pyhap category constants

* Changed 'set_broker' to 'set_driver'

* Changed loader method names

* Use 'serv.configure_char'

* Use 'self.set_info_service'

* Use 'self.add_preload_service'

* Fix hound issue

* Updated HAP-python to 2.0.0

fabaff committed:
Upgrade holidays to 0.9.5 (#14274)

cdce8p committed:
Homekit: Changed device_class requirement Humidity Sensor (#14277)

schmittx committed:
HomeKit: Support triggered state for alarm_control_panel (#14285)

dgomes committed:
Fix filter sensor missing window_size argument (#14252)

* missing window_size argument

* test throttle filter configuration

fabaff committed:
Upgrade astral to 1.6.1 (#14297)

nickw444 committed:
BOM Weather throttle fix (#14234)

blackwind committed:
Publish attributes unconditionally (#14179)

* Publish attributes unconditionally

Because the attribute publish command was previously hidden behind `if val:`, falsy values like False and 0.0 weren't being published, thereby making Statestream -- particularly in the case of booleans, where the first True would be retained indefinitely -- a completely worthless indicator of state.

* Change bool test to False to confirm falsy values pass

cdce8p committed:
Sensor device classes (#14282)

* Added light device class, moved device classes to const

* Removed unnecessary icons

* Replace 'lux' with 'lx'

* Fix comment

* Changed device_class name

tinloaf committed:
Matrix Chatbot (#13355)

* Add first version of the Matrix bot

* It's a stupid but necessary change…

* Dont list it twice

* All hail the linter!

* More linter-pleasing

* Use the correct user ID

* Add expression commands

* Add tests for new validators

* Fix room alias handling

* Wording

* Defer setup

* Simplify commands

* Handle exceptions

* Update requirements

* Review

* Move login back to constructor

* Fix review comments

Kane610 committed:
deCONZ add new device without restart (#14221)

* Add new device without restarting hass

* Remove debug prints

* Fix copy paste error

* Fix comments from balloob
Add tests to verify signalling with new added devices

* Fix hound comments
Add test to verify when new sensor is added

* Fix tests

* Unload entry should unsubscribe all deconz dispatchers

* Make sure mock setup also creates unsub in hass data

* Fix copy paste issue

* Lint

Commit:
Add alarmdotcom sensor status (#14254)

* bump to match Xorso/pyalarmdotcom#9

* Load additional status attributes

* missed a spot

Commit:
Add domain to labels and count state changes to Prometheus (#14253)

* Add domain to labels

* Count state changes

relvacode committed:
Add support for max_volume (#13822)

* onkyo: add support for max volume range

* onkyo: make flake8 happy

* onkyo: fix PEP8 D205 on line 181

* onkyo: use range for max_volume configuration

* onkyo: fix line too long

robmarkcole committed:
Refactor ImageProcessingFaceEntity (#14296)

* Refactor ImageProcessingFaceEntity

* Replace STATE_UNKNOWN with None

robmarkcole committed:
Merge remote-tracking branch 'origin/facebox-detect' into facebox-detect

rsmeral committed:
Onkyo: SUPPORT_VOLUME_STEP (#14299)

teharris1 committed:
Add All-Linking capabilities (#14065)

* Setup all-linking service

* Remove extra line

* Remove linefeed and tab escape chars

* Add services delete_all_link, load_all_link_database and print_all_link_database

* Check if reload is set

* Confirm entity is InsteonPLMEntity before attempting to load or print ALDB

* Debug load and print ALDB

* Debug print aldb

* Debug print_aldb

* Get entity via platform

* Track Insteon entities in component

* Store entity list in hass.data

* Add entity to hass.data

* Add ref to hass in InsteonPLMEntity

* Pass hass correctly to InsteonPLMBinarySensor

* Fix reference to ALDBStatus.PARTIAL

* Print ALDB record as string

* Get ALDB record from memory address

* Reformat ALDB log output

* Add print_im_aldb service

* Remove reference to self in print_aldb_to_log

* Remove reference to self in print_aldb_to_log

* Fix spelling issue with load_all_link_database service

* Bump insteonplm to 0.9.1

* Changes from code review

* Code review changes

* Fix syntax error

* Correct reference to cv.boolean and update requirements

* Update requirements

* Fix flake8 errors

* Reload as boolean test

* Remove hass from entity init

balloob committed:
Update translations

balloob committed:
Update frontend to 20180505.0

balloob committed:
Merge branch 'master' into dev

balloob committed:
Version bump to 0.70.0.dev0

robmarkcole committed:
Address rebase

robmarkcole committed:
Merge remote-tracking branch 'origin/facebox-detect' into facebox-detect

# Conflicts:
#	homeassistant/components/image_processing/facebox_face_detect.py

MartinHjelmare commented:
I often use the `git rebase --onto` command to make sure my branch is rebased on top of the tip of the target branch.

```
git rebase --onto dev LAST_COMMIT_BEFORE_YOUR_BRANCH YOUR_BRANCH
```

robmarkcole commented:
Thanks for the advice @MartinHjelmare 
Is there a way to repair this branch?

MartinHjelmare commented:
Make sure you backup your changed code somehow first. Then you can try this:
https://stackoverflow.com/questions/134882/undoing-a-git-rebase

Disclaimer: I haven't tried it myself.

robmarkcole commented:
This branch is a mess so I'm going to create a fresh one & PR
